File: algorithms/687-longest-univalue-path/longest-univalue-path.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def longest_uni_value_path(root: Optional[TreeNode]) -> int:
    ans = 0

    def dfs(node: Optional[TreeNode]) -> int:
        if node is None:
            return 0
        left = dfs(node.left)
        right = dfs(node.right)
        left_path = left + 1 if node.left and node.left.val == node.val else 0
        right_path = right + 1 if node.right and node.right.val == node.val else 0

        nonlocal ans
        ans = max(ans, left_path + right_path)
        return max(left_path, right_path)

    dfs(root)
    return ans

File: algorithms/687-longest-univalue-path/longest-univalue-path.go
package problem687

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func longestUnivaluePath(root *TreeNode) (ans int) {
	var dfs func(*TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return 0
		}

		left := dfs(node.Left)
		right := dfs(node.Right)
		leftPath := 0
		if node.Left != nil && node.Left.Val == node.Val {
			leftPath = left + 1
		}
		rightPath := 0
		if node.Right != nil && node.Right.Val == node.Val {
			rightPath = right + 1
		}

		ans = max(ans, leftPath+rightPath)
		return max(leftPath, rightPath)
	}

	dfs(root)
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
